src/python/pants/source/source_root.py

```diff
@@ -202,31 +229,55 @@ class OptionalSourceRoot:
 
 
 @rule
-def get_source_root(
+async def get_source_root(
     source_root_request: SourceRootRequest, source_root_config: SourceRootConfig
 ) -> OptionalSourceRoot:
     """Rule to request a SourceRoot that may not exist."""
-    matched_path = source_root_config.get_pattern_matcher().find_root(source_root_request.path)
-    if matched_path:
-        return OptionalSourceRoot(SourceRoot(path=str(matched_path)))
-    else:
-        return OptionalSourceRoot(None)
+    pattern_matcher = source_root_config.get_pattern_matcher()
+    path = source_root_request.path
+
+    # Check if the requested path itself is a source root.
+
+    # A) Does it match a pattern?
+    if pattern_matcher.matches_root_patterns(path):
+        return OptionalSourceRoot(SourceRoot(str(path)))
+
+    # B) Does it contain a marker file?
+    marker_filenames = source_root_config.options.marker_filenames
+    if marker_filenames:
+        for marker_filename in marker_filenames:
+            if (
+                os.path.basename(marker_filename) != marker_filename
```

Comment:
since we are using pathlib here... we might as well avoid os.path...
and use the pathlib equivalent... https://docs.python.org/3/library/pathlib.html#correspondence-to-tools-in-the-os-module
<img width="897" alt="Screenshot 2020-06-05 22 39 50" src="https://user-images.githubusercontent.com/1268088/83936979-9291ca00-a77d-11ea-8a23-2d02557c17ef.png">

Reply:
These are strings though, not paths. In fact this whole check is to verify that they don't contain any path separators. It seems like overkill to convert them to PurePath objects.